File: packages/android-engine/android-js-engine/src/main/java/io/ionic/android_js_engine/JSResponse.kt
package io.ionic.android_js_engine

import org.json.JSONObject

class JSResponse(statusCode: Int, url: String, data: ByteArray?, error: String?) {
    public var ok: Boolean
    public var status: Int
    public var url: String
    public var error: String?

    private var data: ByteArray?

    init {
        this.status = statusCode
        this.ok = statusCode in 201..299
        this.url = url
        this.data = data
        this.error = error
    }

    fun blob(): ByteArray? {
        return data
    }

    fun text(): String {
        val data = this.data ?: return ""
        return data.toString()
    }

    fun json(): JSONObject? {
        val data = this.data ?: return null

        return JSONObject(data.toString());
    }
}

File: packages/capacitor-plugin/ios/Plugin/RunnerEngine/JSResponse.swift
import JavaScriptCore

@objc protocol JSResponseExports: JSExport {
    // swiftlint:disable:next identifier_name
    var ok: Bool { get }
    var status: Int { get }
    var url: String { get }

    func text() -> JSValue
    func json() -> JSValue
}

@objc public class JSResponse: NSObject, JSResponseExports {
    dynamic var url: String
    dynamic var status: Int
    // swiftlint:disable:next identifier_name
    dynamic var ok: Bool {
        return status >= 200 && status <= 299
    }

    private let data: Data?

    public init(status: Int, responseData: Data?) {
        self.status = status
        self.data = responseData
        self.url = ""
    }

    public init(from httpResponse: HTTPURLResponse, responseData: Data?) {
        self.data = responseData
        self.status = httpResponse.statusCode
        self.url = httpResponse.url?.absoluteString ?? ""
    }

    func text() -> JSValue {
        return JSValue(newPromiseIn: JSContext.current()) { resolve, _ in
            guard let data = self.data else {
                resolve?.call(withArguments: [""])
                return
            }

            if let str = String(bytes: data, encoding: .utf8) {
                resolve?.call(withArguments: [str])
                return
            }

            resolve?.call(withArguments: [""])
        }
    }

    func json() -> JSValue {
        return JSValue(newPromiseIn: JSContext.current()) { resolve, reject in
            do {
                guard let data = self.data else {
                    resolve?.call(withArguments: [JSValue(nullIn: JSContext.current()) as Any])
                    return
                }

                let anyObj = try JSONSerialization.jsonObject(with: data)
                let jsonObj = JSValue(object: anyObj, in: JSContext.current())
                resolve?.call(withArguments: [jsonObj as Any])

            } catch {
                let err = JSError(message: error.localizedDescription)
                reject?.call(withArguments: [err as Any])
            }
        }
    }
}
